File: src/wrap-anthropic.ts
import type { Message } from '@anthropic-ai/sdk/resources/messages';
import type { RawMessageStreamEvent } from '@anthropic-ai/sdk/resources/messages';
import { WatchOptions, LogPayload } from './types';
import { calcCost } from './cost';
import { sendLog } from './ingest';

function extractPrompt(params: Record<string, unknown>): string {
  const parts: string[] = [];
  if (typeof params.system === 'string') {
    parts.push(`[system]: ${params.system}`);
  }
  for (const msg of (params.messages as any[]) ?? []) {
    const content =
      typeof msg.content === 'string'
        ? msg.content
        : (msg.content as any[]).map((c: any) => c.text ?? '').join('');
    parts.push(`[${msg.role}]: ${content}`);
  }
  return parts.join('\n');
}

function extractCompletion(response: Message): string {
  return response.content
    .filter(b => b.type === 'text')
    .map(b => (b as { type: 'text'; text: string }).text)
    .join('');
}

function errorPayload(
  start: number,
  model: string,
  prompt: string,
  err: unknown,
): LogPayload {
  return {
    timestamp:     new Date(start).toISOString(),
    model,
    prompt,
    completion:    '',
    latency_ms:    Date.now() - start,
    tokens_used:   0,
    cost_usd:      0,
    error_type:    err instanceof Error ? err.constructor.name : 'UnknownError',
    error_message: err instanceof Error ? err.message : String(err),
  };
}

async function* interceptStream(
  stream: AsyncIterable<RawMessageStreamEvent>,
  start: number,
  params: Record<string, unknown>,
  options: WatchOptions,
): AsyncGenerator<RawMessageStreamEvent> {
  let completion = '';
  let inputTokens = 0;
  let outputTokens = 0;
  let thrownError: unknown;

  try {
    for await (const event of stream) {
      yield event;
      const e = event as any;
      if (e.type === 'message_start') {
        inputTokens = e.message?.usage?.input_tokens ?? 0;
      } else if (e.type === 'content_block_delta' && e.delta?.type === 'text_delta') {
        completion += e.delta.text ?? '';
      } else if (e.type === 'message_delta') {
        outputTokens = e.usage?.output_tokens ?? 0;
      }
    }
  } catch (err) {
    thrownError = err;
    throw err;
  } finally {
    const model = params.model as string;
    if (thrownError !== undefined) {
      sendLog(errorPayload(start, model, extractPrompt(params), thrownError), options);
    } else {
      sendLog({
        timestamp:   new Date(start).toISOString(),
        model,
        prompt:      extractPrompt(params),
        completion,
        latency_ms:  Date.now() - start,
        tokens_used: inputTokens + outputTokens,
        cost_usd:    calcCost(model, inputTokens, outputTokens),
      }, options);
    }
  }
}

/**
 * Patches client.messages.create and client.messages.stream in-place.
 * Returns the same client so callers can chain: const ai = wrapAnthropic(new Anthropic(), opts)
 */
export function wrapAnthropic<T extends { messages: { create: (...args: any[]) => any } }>(
  client: T,
  options: WatchOptions,
): T {
  const orig = client.messages.create.bind(client.messages);

  (client.messages as any).create = async function (
    params: Record<string, unknown>,
    requestOptions?: unknown,
  ): Promise<unknown> {
    const start = Date.now();

    if (params.stream) {
      try {
        const stream = await orig(params, requestOptions);
        return interceptStream(stream, start, params, options);
      } catch (err) {
        sendLog(errorPayload(start, params.model as string, extractPrompt(params), err), options);
        throw err;
      }
    }

    try {
      const response: Message = await orig(params, requestOptions);
      const inputTokens  = response.usage?.input_tokens  ?? 0;
      const outputTokens = response.usage?.output_tokens ?? 0;

      sendLog({
        timestamp:   new Date(start).toISOString(),
        model:       response.model,
        prompt:      extractPrompt(params),
        completion:  extractCompletion(response),
        latency_ms:  Date.now() - start,
        tokens_used: inputTokens + outputTokens,
        cost_usd:    calcCost(response.model, inputTokens, outputTokens),
      }, options);

      return response;
    } catch (err) {
      sendLog(errorPayload(start, params.model as string, extractPrompt(params), err), options);
      throw err;
    }
  };

  // messages.stream() is the high-level streaming helper — it returns a MessageStream
  // (an EventEmitter + AsyncIterable) rather than a raw SSE stream.  We listen for the
  // 'finalMessage' event, which fires once with the fully-accumulated Message after the
  // stream ends, so we get exact token counts without touching the stream data flow.
  if (typeof (client.messages as any).stream === 'function') {
    const origStream = (client.messages as any).stream.bind(client.messages);

    (client.messages as any).stream = function (
      params: Record<string, unknown>,
      requestOptions?: unknown,
    ) {
      const start = Date.now();
      const messageStream = origStream(params, requestOptions);

      messageStream.on('finalMessage', (message: any) => {
        const inputTokens: number  = message.usage?.input_tokens  ?? 0;
        const outputTokens: number = message.usage?.output_tokens ?? 0;
        sendLog({
          timestamp:   new Date(start).toISOString(),
          model:       message.model as string,
          prompt:      extractPrompt(params),
          completion:  extractCompletion(message as Message),
          latency_ms:  Date.now() - start,
          tokens_used: inputTokens + outputTokens,
          cost_usd:    calcCost(message.model as string, inputTokens, outputTokens),
        }, options);
      });

      messageStream.on('error', (err: Error) => {
        sendLog(errorPayload(start, params.model as string, extractPrompt(params), err), options);
      });

      return messageStream;
    };
  }

  return client;
}

File: python-sdk/loglens/_wrap_anthropic.py
import datetime
import time
from typing import Any, Iterator

from ._cost import calc_cost
from ._ingest import send_log
from ._types import LogPayload, WatchOptions


def _extract_prompt(kwargs: dict) -> str:
    parts = []
    for msg in kwargs.get('messages', []):
        content = msg.get('content', '')
        if isinstance(content, str):
            parts.append(content)
        elif isinstance(content, list):
            for block in content:
                if isinstance(block, dict) and block.get('type') == 'text':
                    parts.append(block.get('text', ''))
    return ' '.join(parts)[:2000]


def _now() -> str:
    return datetime.datetime.utcnow().isoformat() + 'Z'


def _error_payload(start: float, model: str, prompt: str, exc: Exception) -> LogPayload:
    return LogPayload(
        timestamp=_now(),
        model=model,
        prompt=prompt,
        completion='',
        latency_ms=int((time.time() - start) * 1000),
        tokens_used=0,
        cost_usd=0.0,
        error_type=type(exc).__name__,
        error_message=str(exc),
    )


class _AnthropicStreamWrapper:
    """Wraps a Stream[MessageStreamEvent] to collect and log after exhaustion."""

    def __init__(self, stream: Any, start: float, model: str, prompt: str, opts: WatchOptions):
        self._stream = stream
        self._start = start
        self._model = model
        self._prompt = prompt
        self._opts = opts

    def __iter__(self) -> Iterator:
        completion_parts = []
        input_tokens = 0
        output_tokens = 0
        thrown = None

        try:
            for event in self._stream:
                event_type = getattr(event, 'type', None)
                if event_type == 'content_block_delta':
                    delta = getattr(event, 'delta', None)
                    if getattr(delta, 'type', None) == 'text_delta':
                        completion_parts.append(getattr(delta, 'text', ''))
                elif event_type == 'message_start':
                    usage = getattr(getattr(event, 'message', None), 'usage', None)
                    if usage:
                        input_tokens = getattr(usage, 'input_tokens', 0)
                elif event_type == 'message_delta':
                    usage = getattr(event, 'usage', None)
                    if usage:
                        output_tokens = getattr(usage, 'output_tokens', 0)
                yield event
        except Exception as exc:
            thrown = exc
            raise
        finally:
            latency = int((time.time() - self._start) * 1000)
            if thrown:
                payload = _error_payload(self._start, self._model, self._prompt, thrown)
                payload.latency_ms = latency
            else:
                total = input_tokens + output_tokens
                payload = LogPayload(
                    timestamp=_now(),
                    model=self._model,
                    prompt=self._prompt,
                    completion=''.join(completion_parts),
                    latency_ms=latency,
                    tokens_used=total,
                    cost_usd=calc_cost(self._model, input_tokens, output_tokens),
                )
            send_log(payload, self._opts)

    def __enter__(self):
        return self

    def __exit__(self, *args):
        if hasattr(self._stream, '__exit__'):
            return self._stream.__exit__(*args)
        return False

    def __getattr__(self, name: str):
        return getattr(self._stream, name)


class _WrappedMessages:
    def __init__(self, messages: Any, opts: WatchOptions):
        self._messages = messages
        self._opts = opts

    def create(self, **kwargs):
        start = time.time()
        model: str = kwargs.get('model', 'unknown')
        prompt = _extract_prompt(kwargs)

        if kwargs.get('stream'):
            raw = self._messages.create(**kwargs)
            return _AnthropicStreamWrapper(raw, start, model, prompt, self._opts)

        try:
            resp = self._messages.create(**kwargs)
            latency = int((time.time() - start) * 1000)

            completion = ''.join(
                getattr(block, 'text', '')
                for block in getattr(resp, 'content', [])
                if getattr(block, 'type', None) == 'text'
            )
            usage = getattr(resp, 'usage', None)
            input_tokens = getattr(usage, 'input_tokens', 0) if usage else 0
            output_tokens = getattr(usage, 'output_tokens', 0) if usage else 0

            payload = LogPayload(
                timestamp=_now(),
                model=model,
                prompt=prompt,
                completion=completion,
                latency_ms=latency,
                tokens_used=input_tokens + output_tokens,
                cost_usd=calc_cost(model, input_tokens, output_tokens),
            )
            send_log(payload, self._opts)
            return resp
        except Exception as exc:
            send_log(_error_payload(start, model, prompt, exc), self._opts)
            raise

    def __getattr__(self, name: str):
        return getattr(self._messages, name)


class WrappedAnthropic:
    def __init__(self, client: Any, opts: WatchOptions):
        self._client = client
        self.messages = _WrappedMessages(client.messages, opts)

    def __getattr__(self, name: str):
        return getattr(self._client, name)


def wrap_anthropic(client: Any, options: WatchOptions) -> WrappedAnthropic:
    return WrappedAnthropic(client, options)
